File: common/src/main/java/guepardoapps/lucahome/common/services/ILucaService.java
package guepardoapps.lucahome.common.services;

import android.content.Context;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.UUID;

public interface ILucaService<T> {
    void Initialize(@NonNull Context context, boolean reloadEnabled, int reloadTimeout, boolean displayNotification, Class<?> receiverActivity);

    void Dispose();

    ArrayList<T> GetDataList() throws NoSuchMethodException;

    T GetByUuid(@NonNull UUID uuid) throws NoSuchMethodException;

    ArrayList<T> SearchDataList(@NonNull String searchKey) throws NoSuchMethodException;

    void LoadData() throws NoSuchMethodException;

    void AddEntry(@NonNull T newEntry) throws NoSuchMethodException;

    void UpdateEntry(@NonNull T updateEntry) throws NoSuchMethodException;

    void DeleteEntry(@NonNull T deleteEntry) throws NoSuchMethodException;

    void SetReloadEnabled(boolean reloadEnabled) throws NoSuchMethodException;

    boolean GetReloadEnabled() throws NoSuchMethodException;

    void SetReloadTimeout(int reloadTimeout) throws NoSuchMethodException;

    int GetReloadTimeout() throws NoSuchMethodException;

    Calendar GetLastUpdate() throws NoSuchMethodException;

    void ShowNotification() throws NoSuchMethodException;

    void CloseNotification() throws NoSuchMethodException;

    void SetDisplayNotification(boolean displayNotification) throws NoSuchMethodException;

    boolean GetDisplayNotification() throws NoSuchMethodException;

    void SetReceiverActivity(Class<?> receiverActivity) throws NoSuchMethodException;

    Class<?> GetReceiverActivity() throws NoSuchMethodException;
}

File: common/src/main/java/guepardoapps/lucahome/common/services/common/ILucaService.kt
package guepardoapps.lucahome.common.services.common

import android.content.Context
import android.support.annotation.NonNull
import guepardoapps.lucahome.common.models.common.RxResponse
import guepardoapps.lucahome.common.models.common.ServiceSettings
import io.reactivex.subjects.PublishSubject
import java.util.*

internal interface ILucaService<T> {
    var initialized: Boolean
    var context: Context?

    var serviceSettings: ServiceSettings
    var receiverActivity: Class<*>?

    val responsePublishSubject: PublishSubject<RxResponse>

    fun initialize(context: Context)
    fun dispose()

    fun get(): MutableList<T>
    fun get(uuid: UUID): T?
    fun search(@NonNull searchValue: String): MutableList<T>

    fun load()

    fun add(@NonNull entry: T, reload: Boolean)
    fun update(@NonNull entry: T, reload: Boolean)
    fun delete(@NonNull entry: T, reload: Boolean)
}